### PR TITLE
initiate project

### PR DESCRIPTION
**Pull request by maykremer**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,17 +1,34 @@
...
+  },
+  "dependencies": {
+    "joi": "^17.3.0",
+    "mongoose": "^5.11.15",
+    "ts": "^0.2.2",
+    "tsc": "^1.20150623.0",
+    "@types/jest": "^26.0.20",
+    "@typescript-eslint/eslint-plugin": "2.31.0",
+    "@typescript-eslint/parser": "2.31.0",
+    "cross-env": "^6.0.3",
```
Comment (vitaly-furman) [inline package.json:21]: I don't thing you use this, remove it.

File: package.json [hunk 2]
```diff
@@ -1,17 +1,34 @@
...
+    "lint": "eslint -c .eslintrc.json --no-error-on-unmatched-pattern ./src/**/*"
+  },
+  "dependencies": {
+    "joi": "^17.3.0",
+    "mongoose": "^5.11.15",
+    "ts": "^0.2.2",
+    "tsc": "^1.20150623.0",
+    "@types/jest": "^26.0.20",
+    "@typescript-eslint/eslint-plugin": "2.31.0",
+    "@typescript-eslint/parser": "2.31.0",
```
Comment (vitaly-furman) [inline package.json:20]: Move to devDependancies.

File: package.json [hunk 3]
```diff
@@ -1,17 +1,34 @@
...
     "start": "func start",
-    "test": "echo \"No tests yet...\""
+    "test": "tsc && jest",
+    "lint": "eslint -c .eslintrc.json --no-error-on-unmatched-pattern ./src/**/*"
+  },
+  "dependencies": {
+    "joi": "^17.3.0",
+    "mongoose": "^5.11.15",
+    "ts": "^0.2.2",
+    "tsc": "^1.20150623.0",
```
Comment (vitaly-furman) [inline package.json:17]: Unused, remove.

**Review by vitaly-furman — CHANGES_REQUESTED** [review]

**Review by vitaly-furman — DISMISSED** [review]

File: events-service/index.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+                nodesGroups.forEach((node) => {
+                    if (!node.interviewers) throw new ApplicationError(404, 'There is no interviewers');
+                    node.interviewers.forEach((interviewerId: mongoose.ObjectId) => {
+                        interviewersList.push(interviewerId);
+                    });
+                });
+            }
+
+            const eventObject = { ...queueObj, interviewersIds: [...interviewersList] };
+            await EventModel.create(eventObject).catch((err) => {
```
Comment (yonatandt) [inline events-service/index.ts:34]: So does this function just create an event? what about updating the event (like the Trello card says)?
@yubbi99

File: events-service/nodesGroup.interface.ts [hunk 5]
```diff
@@ -0,0 +1,8 @@
+import { ObjectId } from 'mongoose';
+
+export default interface INodesGroup {
+    name: string;
+    owners: [ObjectId];
+    interviewers: [ObjectId];
+    nodes: [string];
+}
```
Comment (yonatandt) [inline events-service/nodesGroup.interface.ts:8]: Shouldn't be in this function

File: events-service/nodesGroup.model.ts [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+const nodesGroupSchema = new mongoose.Schema({
+    name: String,
+    owners: [mongoose.Types.ObjectId],
+    interviewers: [mongoose.Types.ObjectId],
+    nodes: [String],
+});
+
+const nodesGroupModel = mongoose.model<INodesGroup & mongoose.Document>(process.env.NODES_GROUP_COLLECTION_NAME , nodesGroupSchema);
+
+export default nodesGroupModel;
```
Comment (yonatandt) [inline events-service/nodesGroup.model.ts:13]: Shouldn't be in this function.

File: events-service/readme.md [hunk 7]
```diff
@@ -0,0 +1,64 @@
+# Event Queue Triggeer
+
+## Queue Storage Triggeer
+
+listen to events-queue
```
Comment (yonatandt) [inline events-service/readme.md:5]: and do what?

File: events-service/sample.dat [hunk 8]
```diff
@@ -0,0 +1 @@
+sample queue data
```
Comment (yonatandt) [inline events-service/sample.dat:1]: Either put a sample or remove it from git.

File: events-service/utils/validate.ts [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+import Status from './status.enum';
+
+const validEvent = Joi.object({
+    nodeId: Joi.string().required(),
+    malshabId: Joi.string().required(),
+    time: Joi.date().required(),
+    location: Joi.string().required(),
+    status: Joi.string().required().valid(Status.Status1, Status.Status2), // TODO: insert statuses
+    url: Joi.string(),
+});
```
Comment (yonatandt) [inline events-service/utils/validate.ts:11]: So are all of these fields required in each message?

File: events-service/utils/validate.ts [hunk 10]
```diff
@@ -0,0 +1,13 @@
+import * as Joi from 'joi';
+import Status from './status.enum';
+
+const validEvent = Joi.object({
+    nodeId: Joi.string().required(),
+    malshabId: Joi.string().required(),
+    time: Joi.date().required(),
+    location: Joi.string().required(),
+    status: Joi.string().required().valid(Status.Status1, Status.Status2), // TODO: insert statuses
```
Comment (yonatandt) [inline events-service/utils/validate.ts:9]: So maybe you shouldn't put it here.
Note that right now, an object will not be valid if the status is not `Status1` or `Status2`.

**Review by yonatandt — CHANGES_REQUESTED** [review]
What is this function purpose? It is not clear to me.
Does it create an Event object, can it update an event object?

**Comment by maykremer** [comment]
> What is this function purpose? It is not clear to me.
> Does it create an Event object, can it update an event object?

Just create. Events API update events.

File: events-service/index.ts (hunk 4 above)
Comment (maykremer) [reply events-service/index.ts:34]: Events api update events

File: events-service/nodesGroup.interface.ts (hunk 5 above)
Comment (maykremer) [reply events-service/nodesGroup.interface.ts:8]: I need it in order to get the interviewer's array

File: events-service/nodesGroup.model.ts (hunk 6 above)
Comment (maykremer) [reply events-service/nodesGroup.model.ts:13]: I need it in order to get the interviewer's array

File: events-service/utils/validate.ts (hunk 9 above)
Comment (maykremer) [reply events-service/utils/validate.ts:11]: yes.

File: events-service/index.ts [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+            .catch((err) => {
+                throw new ApplicationError(err.message);
+            });
+        const interviewersList: Array<mongoose.Types.ObjectId> = [];
+        if (nodesGroups) {
+            nodesGroups.forEach((node) => {
+                if (!node.interviewers) throw new ApplicationError('There is no interviewers');
+                node.interviewers.forEach((interviewerId: mongoose.Types.ObjectId) => {
+                    interviewersList.push(interviewerId);
+                });
```
Comment (carmeltsadok) [inline events-service/index.ts:29]: I think you should check if the interviewer already exists in the array so there won't be repeated ids.

File: events-service/index.ts [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+        await getConnection();
+        const nodesGroups = await NodesGroupModel.find({ nodes: queueObj.nodeId })
+            .exec()
+            .catch((err) => {
+                throw new ApplicationError(err.message);
+            });
+        const interviewersList: Array<mongoose.Types.ObjectId> = [];
+        if (nodesGroups) {
+            nodesGroups.forEach((node) => {
+                if (!node.interviewers) throw new ApplicationError('There is no interviewers');
```
Comment (carmeltsadok) [inline events-service/index.ts:26]: interviewers is a required field in nodesGroup, so why it won't exist?

File: events-service/index.ts [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+        }
+        await getConnection();
+        const nodesGroups = await NodesGroupModel.find({ nodes: queueObj.nodeId })
+            .exec()
+            .catch((err) => {
+                throw new ApplicationError(err.message);
+            });
+        const interviewersList: Array<mongoose.Types.ObjectId> = [];
+        if (nodesGroups) {
+            nodesGroups.forEach((node) => {
```
Comment (carmeltsadok) [inline events-service/index.ts:25]:
```suggestion
            nodesGroups.forEach((nodesGroup) => {

```

File: events-service/function.json [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+    {
+      "name": "myQueueItem",
+      "type": "queueTrigger",
+      "direction": "in",
+      "queueName": "events-queue",
+      "connection": "AzureWebJobsStorage"
+
+    }
+  ],
+  "scriptFile": "../dist/events-service/index.js"
```
Comment (carmeltsadok) [inline events-service/function.json:12]:
```suggestion
  "scriptFile": "../dist/CreateEvent/index.js"
```
I think it should have a meaningful name and according to the naming conventions.

File: events-service/index.ts [hunk 15]
```diff
@@ -0,0 +1,41 @@
+import { AzureFunction, Context } from '@azure/functions';
+import * as mongoose from 'mongoose';
+import getConnection from '../shared/utils/db';
+import EventModel from '../shared/models/event.model';
+import NodesGroupModel from '../shared/models/nodesGroup.model';
+import ApplicationError from './utils/error';
+import validEvent from './utils/validate';
+import { IEvent } from '../shared/interfaces/event.interface';
+
+const createEvent: AzureFunction = async (context: Context, myQueueItem: IEvent): Promise<void> => {
```
Comment (carmeltsadok) [inline events-service/index.ts:10]: Give a meaningful name to the queueItem.

**Review by carmeltsadok — CHANGES_REQUESTED** [review]

**Pull request closed without merging**